use GH actions insteaad of travis for CI

dylanjha opened the pull request:
TODO:

- [x] use actions to update docs when merged to master
- [x] use actions to publish new versions to npm
- [x] right now all our changes are in `CHANGELOG.md`, I think we should go edit the github releases and move all of those notes into the Github Releases panel and stop using CHANGELOG.md. Moving forward, creating a new release in Github will trigger an npm publish, which is a better dev workflow

I have no idea if the npm publish or update docs workflow actually **works**, sort of yolo'ing that but would love a way to be more sure of those changes before merging.

dylanjha commented on .github/workflows/npm-publish.yml:
```diff
@@ -0,0 +1,32 @@
+      - run: yarn install --frozen-lockfile
+      - run: yarn publish
+        env:
+          NODE_AUTH_TOKEN: ${{secrets.npm_token}}
```
@mmcc is this npm_token stored at the Github org level? Or does it have to be added to each repo?

I riffed off your work here: https://github.com/muxinc/upchunk/blob/master/.github/workflows/npm-publish.yml

mmcc replied:
Just to follow up on this so it's in-line, yes, the npm token is now stored at the org level.

dylanjha commented on .github/workflows/update-docs.yml:
```diff
@@ -0,0 +1,26 @@
+      - name: Deploy Docs
+        uses: JamesIves/github-pages-deploy-action@3.7.1
+        with:
+          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
```
I *think* this secret might already be available?

dylanjha commented:
@jsanford8 @mmcc @GeneticGenesis I want to get this wrapped up. 

@mmcc added NPM_TOKEN into org secrets, so it should be available in this workflow `secrets.NPM_TOKEN`. Also, based on [this guide](https://docs.github.com/en/free-pro-team@latest/actions/reference/authentication-in-a-workflow#about-the-github_token-secret) it sounds like `GITHUB_TOKEN` will also be available.

I think my best bet here is to:

1. merge this
2. create a new branch that bumps `3.1.1` with some small recent bugfixes that have been merged
3. follow the steps in README_INTERNAL.md (tag master, push tag, create a release in the Github UI)
4. 🤞 that it works, if it doesn't, then debug, fix the workflow and try again?

I wish I could test (3) better so that I can avoid what will probably happen: a few follow up PRs to get the workflow working right, but I can't really think of a way to test that without actually _doing it_.

mmcc commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,22 @@
+    - run: yarn install --frozen-lockfile
+    - run: yarn format
+    - run: yarn test:unit
+    - run: yarn test:int
```
I could be wrong, but shouldn't just `yarn test` work? It should recursively run everything in the `test` directory, and we don't want it doing things like recording _new_ Nock VCRs in this environment anyway.

dylanjha replied:
Hmmm,

Let me see. `yarn test:int` will run `NOCK_BACK_MODE=record mocha --recursive ./test/integration --timeout 10000`

To review what that means:

* `record`: use recorded nocks, record new nocks

Other options:

* `wild`: all requests go out to the internet, don't replay anything, doesn't record anything
* `dryrun`: The default, use recorded nocks, allow http calls, doesn't record anything, useful for writing new tests
* `record`: use recorded nocks, record new nocks
* `lockdown`: use recorded nocks, disables all http calls even when not nocked, doesn't record

Looking at these definitions, I think you're right. `record` is what we want when developing locally (it will use recorded tests if they're there, otherwise record new ones)

For CI the default makes the most sense. Maybe in test we want to run `NOCK_BACK_MODE=lockdown yarn test`? That's the same idea you had, except setting the Nock mode to lockdown so that we don't accidentally make unintended http requests.

mmcc replied:
I can get behind that. The default (dryrun) _shouldn't_ make any live requests since the only requests should be `nocked` anyway, but it's not the worst idea just in case.

dylanjha replied:
I had the same thought... okay, committing!

mmcc commented on .github/workflows/npm-publish.yml:
```diff
@@ -0,0 +1,32 @@
+      - run: yarn install --frozen-lockfile
+      - run: yarn format
+      - run: yarn test:unit
+      - run: yarn test:int
```
Same comment as above, I think in this environment we just want to use `yarn test`

mmcc commented on README_INTERNAL.md:
```diff
@@ -1,9 +1,6 @@
-1. Update CHANGELOG.md
 1. Update version in package.json
 1. Open Pull Request, merge after approved
 1. Tag master: example: `git tag v2.6.0` and `git push --tags origin`
```
Should we be suggesting doing this directly or using `npm version` to both bump package.json and create the tag?

dylanjha replied:
good shout, I like that better!

mmcc reviewed (approved):
This looks good to me, unfortunately you're totally right that the sucky thing here is just neededing to hope and pray at some point when it comes to the first big Github Action release. There was an errant version of UpChunk released because of this, so just keep an eye on the first time any of these that update other systems run.

dylanjha commented:
@mmcc I believe I addressed your feedback. Thanks for that review! I confirmed that `NOCK_BACK_MODE=lockdown yarn test` works as expected (runs all int & unit tests in this action run): https://github.com/muxinc/mux-node-sdk/pull/85/checks?check_run_id=1693172267

Does my updated README_INTERNAL.md steps look right now?

mmcc commented on README_INTERNAL.md:
```diff
@@ -1,9 +1,6 @@
-1. Run `npm publish`
-
-If releaseing a beta tag, run `npm publish --tag beta`
+1. Create a new release in the Github UI, give the release a name and add release notes (creating the release will kick off npm publish)
```
The only addition I might make here would be to say that you need to make sure to select the tag created in the previous step when creating the new release, but otherwise this looks great! Do we want to include a template of what a release body should look like in here?

dylanjha replied:
Good idea, I can include a template that when I go through the next release (a few small bug fixes)

mmcc reviewed (approved):
This looks fantastic! 🚢
